libraries/AP_Hygrometer/AP_Hygrometer_UAVCAN.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Hygrometer.h"
+
+#if AP_HYGROMETER_ENABLED
+
+#if HAL_ENABLE_LIBUAVCAN_DRIVERS
+#include <AP_HAL/AP_HAL.h>
+#include "AP_Hygrometer_UAVCAN.h"
+#include <dronecan/sensors/hygrometer/Hygrometer.hpp>
+#include <AP_CANManager/AP_CANManager.h>
+
+#define LOG_TAG "Hygrometer"
+#define C_TO_KELVIN 273.15f
+
+UC_REGISTRY_BINDER(HygrometerCb, dronecan::sensors::hygrometer::Hygrometer);
+
+AP_Hygrometer_UAVCAN::DetectedModules AP_Hygrometer_UAVCAN::_detected_modules[] = {0};
+HAL_Semaphore AP_Hygrometer_UAVCAN::_sem_registry;
+
+AP_Hygrometer_UAVCAN::AP_Hygrometer_UAVCAN(AP_Hygrometer::Param &_params) :
+    AP_Hygrometer_Backend(_params)
+{}
+
+void AP_Hygrometer_UAVCAN::subscribe_msgs(AP_UAVCAN* ap_uavcan)
+{
+    if (ap_uavcan == nullptr) {
+        return;
+    }
+
+    auto* node = ap_uavcan->get_node();
+
+    uavcan::Subscriber<dronecan::sensors::hygrometer::Hygrometer, HygrometerCb> *hygrometer_listener;
+    hygrometer_listener = new uavcan::Subscriber<dronecan::sensors::hygrometer::Hygrometer, HygrometerCb>(*node);
+
+    const int hygrometer_listener_res = hygrometer_listener->start(HygrometerCb(ap_uavcan, &handle_hygrometer));
+
+    if (hygrometer_listener_res < 0) {
+        AP_HAL::panic("UAVCAN hygrometer subscriber start problem\n");
+    }
+}
+
+AP_Hygrometer_Backend* AP_Hygrometer_UAVCAN::probe(AP_Hygrometer::Param &_params)
+{
+    WITH_SEMAPHORE(_sem_registry);
+
+    AP_Hygrometer_UAVCAN* backend = nullptr;
+
+    for (uint8_t i = 0; i < AP_HYGROMETER_MAX_SENSORS; i++) {
+        if (_detected_modules[i].driver == nullptr && _detected_modules[i].ap_uavcan != nullptr) {
+            backend = new AP_Hygrometer_UAVCAN(_params);
+            if (backend == nullptr) {
+                AP::can().log_text(AP_CANManager::LOG_INFO,
+                                   LOG_TAG,
+                                   "Failed register UAVCAN Hygrometer Node %d on Bus %d\n",
```

Comment:
This doesn't look right as an error message; `new` will always return an object even if the thing fails to register.